Machine_Learning/notebooks/czmodel/SingleClassSemanticSegmentation_PyTorch_5_0_0.ipynb

```diff
@@ -0,0 +1,684 @@
+    "\n",
+    "        return img.to(torch.float32), msk.to(torch.float32)\n",
+    "\n",
+    "    def __len__(self):\n",
+    "        return len(self.sample_images)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Define transforms\n",
+    "from torchvision import transforms\n",
+    "transforms = transforms.Compose([\n",
+    "    transforms.ToTensor(),\n",
+    "    transforms.CenterCrop(CROP_SIZE),\n",
+    "])\n",
+    "\n",
+    "# Create dataset and dataloader\n",
+    "sample_dataset = SampleDataset(images_folder=IMAGES_FOLDER,\n",
+    "                               masks_folder=MASKS_FOLDER,\n",
+    "                               transforms=transforms)\n",
+    "\n",
+    "sample_dataloader = DataLoader(dataset=sample_dataset,\n",
+    "                                      batch_size=2,\n",
+    "                                      shuffle=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "KcumQa0yAsxb",
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "### Define a simple model\n",
+    "This part defines a PyTorch model with two convolutional layers and softmax activation at the output node. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "jW6eR6s8Asxc",
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# Define simple PyTorch model with two convolutional layers and softmax activation at the output node\n",
+    "class SampleModel(nn.Module):\n",
+    "\n",
+    "    def __init__(self):\n",
+    "        super(SampleModel, self).__init__()\n",
+    "        self.conv = nn.Sequential(\n",
+    "            nn.Conv2d(3, 16, 3, padding=1),  \n",
+    "            nn.Conv2d(16, 2, 1, padding=0),\n",
+    "            nn.Softmax(),\n",
+    "        )\n",
+    "\n",
+    "    def forward(self, x):\n",
+    "        x = self.conv(x)\n",
+    "        return x\n",
+    "\n",
+    "\n",
+    "model = SampleModel().to(device)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "MpBMYhGzAsxc",
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "### Training the model with the loaded data\n",
+    "This part fits the model to the loaded data. In this test example we do not care about an actual evaluation of the model using validation and test datasets."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "jSSU22TtAsxd",
+    "outputId": "7854dc6c-3fc4-426c-a5c3-6328b0c5f4c8",
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# Define the number of training epochs\n",
+    "NUM_EPOCHS = 5\n",
+    "\n",
+    "# Define loss function and optimizer\n",
+    "loss_fn = torch.nn.BCELoss()\n",
```

Comment:
Does this actually work? I would expect that BCE loss requires one channel (representing the probability of the pixel to be of class Nucleus).
Since you are using a SoftMax activation where you get a proability distribution over the classes `Background` and `Nucleus` it feels like the categorical crossentropy is what you are looking for: https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html.